### PR TITLE
Cereal integration

### PR DESCRIPTION
**Pull request by Ragingram2**
pls

### CODE REVIEW

File: applications/sandbox/source.cpp [hunk 1]
```diff
@@ -12,6 +12,9 @@
 
 #include "module/testModule.hpp"
 
+#include <core/scenemanagement/SceneManagerModule.hpp>
+#include <core/scenemanagement/scene.hpp>
```
Comment (GlynLeine) [inline applications/sandbox/source.cpp]: maybe merge the scene manager module with the core module, that already gets automatically reported. Scenes are part of core.

File: applications/sandbox/systems/testsystem.hpp [hunk 2]
```diff
@@ -111,6 +112,9 @@ class TestSystem final : public System<TestSystem>
             delete[] buf;
         }
 
+        std::cout << "Yay";
+        std::cout << "please run";
```
Comment (GlynLeine) [inline applications/sandbox/systems/testsystem.hpp]: we have a logging system

File: applications/test_game/systems/testsystem.hpp [hunk 3]
```diff
@@ -0,0 +1,754 @@
...
+         }*/
+
+         //if (accumulated > 10.f)
+         //{
+         //	std::cout << "raising exit event" << std::endl;
+         //	raiseEvent<events::exit>();
+         //	//throw args_exception_msg("hehehe fuck you >:D");
+         //}
+    }
+};
```
Comment (GlynLeine) [inline applications/test_game/systems/testsystem.hpp]: this file shouldn't exist

File: legion/engine/core/defaults/coremodule.hpp [hunk 4]
```diff
@@ -30,6 +31,8 @@ namespace legion::core
             reportComponentType<position>();
             reportComponentType<rotation>();
             reportComponentType<scale>();
+            reportComponentType<scenemanagement::scene>();
+            reportSystem < ::legion::core::scenemanagement::SceneManager > ();
```
Comment (GlynLeine) [inline legion/engine/core/defaults/coremodule.hpp]: wait you already report in core module, so why does scene module exist?

File: legion/engine/core/ecs/component_handle.hpp [hunk 5]
```diff
@@ -72,7 +77,8 @@ namespace legion::core::ecs
 
         bool operator==(const component_handle<component_type>& other) const { return m_registry == other.m_registry && m_ownerId == other.m_ownerId; }
 
-        /**@brief Thread-safe read of component.
+        /**@brief Atomic read of component.
+         * @param order Memory order at which to load the component.
```
Comment (GlynLeine) [inline legion/engine/core/ecs/component_handle.hpp]: please return my comments to what they used to be

File: legion/engine/core/ecs/component_handle.hpp [hunk 6]
```diff
@@ -110,10 +116,10 @@ namespace legion::core::ecs
...
         }
 
-        /**@brief Thread-safe write of component.
-         * @param value Value you wish to write.
-         * @returns component_type Current value of component.
-         */
+        ///**@brief Thread-safe write of component.
+        // * @param value Value you wish to write.
+        // * @returns component_type Current value of component.
+        // */
```
Comment (GlynLeine) [inline legion/engine/core/ecs/component_handle.hpp]: please don't comment out my comments

File: legion/engine/core/ecs/entity_handle.hpp [hunk 7]
```diff
@@ -291,9 +303,12 @@ namespace legion::core::ecs
...
          */
         template<typename component_type>
-        component_handle<std::remove_reference_t<component_type>> add_component(component_type&& value) const
+        component_handle<std::remove_reference_t<component_type>> add_component(component_type&& value, std::memory_order order = std::memory_order_release) const
         {
-            return force_value_cast<component_handle<std::remove_reference_t<component_type>>>(add_component(typeHash<std::remove_reference_t<component_type>>(), reinterpret_cast<void*>(&value)));
+            using actual_type = std::remove_reference_t<component_type>;
+            component_handle<actual_type> handle = force_value_cast<component_handle<actual_type>>(add_component(typeHash<actual_type>()));
+            handle.write(value);
+            return handle;
```
Comment (GlynLeine) [inline legion/engine/core/ecs/entity_handle.hpp]: please return this function to what it should be

File: legion/engine/core/engine/system.hpp [hunk 8]
```diff
@@ -47,7 +47,7 @@ namespace legion::core
 
             m_scheduler->hookProcess<charc>(processChainName, m_processes[id].get());
         }
-
+        /*
```
Comment (GlynLeine) [inline legion/engine/core/engine/system.hpp]: what's this?

File: legion/engine/core/scenemanagement/scenemanager.cpp [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+
+    void SceneManager::createScene(const std::string& name)
+    {
+        scene s;
+        s.id = nameHash(name);
+        sceneNames.emplace(s.id, name);
+        auto sceneHandle = registry->createEntity().add_component(s);
+        sceneList.emplace(nameHash(name), sceneHandle);
+    }
```
Comment (GlynLeine) [inline legion/engine/core/scenemanagement/scenemanager.cpp]: maybe add an overload that allows you to pass an already existing entity

File: legion/engine/core/scenemanagement/scenemanager.hpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        int sceneCount;
+        static std::unordered_map < id_type, std::string> sceneNames;
+        static std::unordered_map < id_type, ecs::component_handle <scene > > sceneList;
+        static legion::core::ecs::EcsRegistry* registry;
+
+        SceneManager() = default;
+
+        virtual void setup()
+        {
+            registry = m_ecs;
```
Comment (GlynLeine) [inline legion/engine/core/scenemanagement/scenemanager.hpp]: shouldn't rly be necessary anymore m_ecs is already static.

File: applications/sandbox/systems/testsystem.hpp (hunk 2 above)
Comment (Ragingram2) [reply applications/sandbox/systems/testsystem.hpp]: Just added it so I could attempt the analyze/build actions to even notice my changes. but thats fixed now

File: legion/engine/core/defaults/coremodule.hpp (hunk 4 above)
Comment (Ragingram2) [reply legion/engine/core/defaults/coremodule.hpp]: scene is a component though

File: legion/engine/core/scenemanagement/scenemanager.cpp (hunk 9 above)
Comment (Ragingram2) [reply legion/engine/core/scenemanagement/scenemanager.cpp]: like a copy method?

**Review by Algo-ryth-mix — APPROVED** [review]

**Review by GlynLeine — APPROVED** [review]

**Pull request merged**